WIP: Make blueprint output pass type-checking

simonihmig opened the pull request:
First step in an attempt to fix #4.

For the beginning this adds a (failing) test to type-check the blueprint output. Actually fixing the output will follow over time... (at least waiting for #74 to be merged) 

The test makes sure that files generated by the blueprints type-check correctly. As all tests use fixture files to test that the blueprints generate the correct output, we can just add a linting test that type-checks the existing static fixtures files, instead of running the blueprints and type-checking the output afterwards.

To make imports not cause type errors, the appropriate typing packages have been added to `devDependencies`, and `types/index.d.ts` declares all in-app modules used in the test fixtures.

Anyone here have any concerns with the approach taken here so far? @chriskrycho @mike-north

chriskrycho commented:
This reply is five months late in coming, mostly because I have not made the time for thinking through the blueprints space. I apologize!

Per the various discussions we’ve had around #74 and #71, I’m unsure how we should proceed. I’d like to talk through this point with some Ember CLI folks to see if we can design a better solution. In the meantime, I’ll leave it at your discretion whether you want to leave this and #74 open, or close them till we make a call (hopefully soon!).

chriskrycho commented:
@simonihmig any interest in rebasing this and landing it, at least as a stopgap while we iterate toward something better? Making sure our output type-checks seems like a big win. 😅

chriskrycho commented on tsconfig.json:
```diff
@@ -0,0 +1,33 @@
+{
```
I like having this in place; let's rename it to `test.tsconfig.json` so that it doesn't get picked up on the editor side. We may want to actually migrate this repo to TS, so we need the test-specific config to be separate!

chriskrycho commented on package.json:
```diff
@@ -15,6 +15,7 @@
   "scripts": {
     "build": "ember build",
     "lint:js": "eslint .",
+    "lint:blueprints": "tsc -noEmit",
```
Per comment on the `tsconfig.json` file, we should reference the test config instead. Also there's a typo here: should be `--noEmit`.

```suggestion
    "lint:blueprints": "tsc test.tsconfig.json --noEmit",
```

tdwesten commented:
@simonihmig please, take a look at the comments form @chriskrycho!

simonihmig commented:
Sorry folks, somehow I missed the previous comments... (probably because they were posted on my birthday, and I had better things to do :wink:)

I just noticed that https://github.com/emberjs/rfcs/pull/776 has just been merged, so we now have a way better path forward to support TS blueprints. Therefore I see little value in investing more time into this. @chriskrycho do you agree?

chriskrycho commented:
@simonihmig I actually think it’s still *very* valuable, and I also have time to shepherd it across the line if you have time to rebase it/get it up to date. The reason it’s still valuable: while we have the ability to author blueprints in TS, we still have some design work to do in terms of what the content of those blueprints should be. That’s relatively small, and I will have the RFC written this quarter—probably in February, but we even if it’s the easiest turn-around ever on an RFC, and we land the blueprints implementations the second it merges, that would put it landing in stable Ember CLI 4.4 in June *at the earliest*. Soooooo if you have time, let’s get this done!

simonihmig commented:
@chriskrycho I rebased this, applied your suggestions, and updated the related dependencies (e.g. latest TS). 

Also added the linting step to CI, which is for now allowed to fail, as there are expectedly a bunch of [typing errors](https://github.com/typed-ember/ember-cli-typescript-blueprints/runs/4748700330?check_suite_focus=true#step:6:6) in the existing blueprint files.

chriskrycho commented:
Excellent. I'll re-review on Monday! Thank you!

simonihmig commented:
> Excellent. I'll re-review on Monday! Thank you!

@chriskrycho ping :wink:

chriskrycho commented:
Just found this while finally (!) clearing out email. Let’s wait till we land #240 to avoid blocking that. Going to try to make that happen early this next week. Thanks for the ping!

muziejus commented:
I can give a preview of what will happen once #240 lands:

- [ ] Errors regarding `store` in ember-data tests. I solved this by doing `let store = this.owner.lookup('service:store') as Store;` and importing the Store in the test file. Not a big lift.
- [ ] However, when I do that, in the `this.store.createRecord()` call in the next line of the test blueprint, I get this error:  `Argument of type 'string' is not assignable to parameter of type 'never'`, as `createRecord()` is expecting: `DS.Store.createRecord<K extends never>(modelName: K, inputProperties?: {}): ModelRegistry[K]`. This is not what I get in my app, where `<K extends ModelRegistry>`.
- [ ] Errors regarding old testing apis, namely `Property 'subject' does not exist on type 'TestContext'` on pre-rfc232 qunit files. In the same vein...
- [ ] `Module '"ember-qunit"' has no exported member 'moduleFor'.` in the old qunit files. I'm inclined to just change these blueprints back to regular old js, if you're doing pre-rfc232 testing…
- [ ] No type given to `params` in the `route-with-dynamic-segment` fixture.
- [ ] Problems loading related models in the model fixtures.
- [ ] A handful of typing problems regarding the `@attr` typing, as noted elsewhere. (`Argument of type '"array"' is not assignable to parameter of type 'keyof TransformRegistry'` etc)
- [ ] issues with `instance` and `application`, etc., in instance initializer tests
- [ ] declared but unused `appInstance` param in intance-initializer fixtures.
- [ ] Can't find `my-app/config/environment` in instance-initializer test.
- [ ] Same for `../../helpers/destroy-app` in instance-initializer test fixtures.
- [ ] Several fixtures have... jQuery? (`assert.strictEqual(this.$().text().trim(), 'template block text')`) Either way, it's not an existing property on `TestContext`.
- [ ] `/node-tests/fixtures/mu-helper.ts` has implicit any for `positional`
- [ ] `assert.dom()` is an unknown property on `assert`
- [ ] `this.element` can possibly be null, but we don't check against that in, for example, `node-tests/fixtures/helper-test/integration.ts`.
- [ ] `Cannot find module '../../tests/helpers/module-for-acceptance' or its corresponding type declarations` in acceptance test fixtures.
- [ ] And `application` variable issues in `node-tests/fixtures/acceptance-test/mocha.ts`
- [ ] Can't find type declarations for `@glimmer/component`
- [ ] Need to type `this.owner.factoryFor()` so that `.create()` is recognized in `node-tests/fixtures/component-test/rfc232-unit-ts`

Here are the complaints:
```
Found 108 errors in 45 files.

Errors  Files
     1  node-tests/fixtures/acceptance-test/addon-default.ts:2
     1  node-tests/fixtures/acceptance-test/addon-nested.ts:2
     2  node-tests/fixtures/acceptance-test/mocha.ts:7
     3  node-tests/fixtures/component-test/default-template.ts:1
     3  node-tests/fixtures/component-test/default.ts:1
     2  node-tests/fixtures/component-test/mocha-rfc232.ts:16
     1  node-tests/fixtures/component-test/rfc232-unit.ts:8
     2  node-tests/fixtures/component-test/rfc232.ts:15
     3  node-tests/fixtures/component-test/unit.ts:1
     1  node-tests/fixtures/component/component-dash.ts:1
     1  node-tests/fixtures/component/component-nested.ts:1
     1  node-tests/fixtures/component/component.ts:1
     2  node-tests/fixtures/controller-test/default-nested.ts:1
     2  node-tests/fixtures/controller-test/default.ts:1
     2  node-tests/fixtures/helper-test/integration.ts:1
     1  node-tests/fixtures/helper-test/mocha-rfc232.ts:16
     1  node-tests/fixtures/helper-test/rfc232.ts:15
     1  node-tests/fixtures/helper/mu-helper.ts:3
     4  node-tests/fixtures/initializer-test/default.ts:10
     4  node-tests/fixtures/initializer-test/dummy.ts:10
     3  node-tests/fixtures/initializer-test/mocha.ts:8
     7  node-tests/fixtures/initializer-test/rfc232.ts:3
     1  node-tests/fixtures/initializer/initializer-nested.ts:3
     1  node-tests/fixtures/initializer/initializer.ts:3
     6  node-tests/fixtures/instance-initializer-test/default.ts:9
     6  node-tests/fixtures/instance-initializer-test/dummy.ts:9
     6  node-tests/fixtures/instance-initializer-test/mocha.ts:6
    10  node-tests/fixtures/instance-initializer-test/rfc232.ts:3
     1  node-tests/fixtures/instance-initializer/instance-initializer-nested.ts:3
     1  node-tests/fixtures/instance-initializer/instance-initializer.ts:3
     1  node-tests/fixtures/model-test/comment-default.ts:10
     1  node-tests/fixtures/model-test/foo-default.ts:10
     2  node-tests/fixtures/model-test/octane-comment-belongs-to.ts:2
     5  node-tests/fixtures/model-test/octane-foo-attrs.ts:2
     1  node-tests/fixtures/model-test/octane-post-has-many.ts:2
     1  node-tests/fixtures/model-test/post-default.ts:10
     2  node-tests/fixtures/route-test/default-child.ts:1
     2  node-tests/fixtures/route-test/default-nested.ts:1
     2  node-tests/fixtures/route-test/default.ts:1
     1  node-tests/fixtures/route/route-with-dynamic-segment.ts:4
     2  node-tests/fixtures/serializer-test/application-default.ts:10
     2  node-tests/fixtures/serializer-test/foo-default.ts:10
     2  node-tests/fixtures/serializer-test/rfc232.ts:10
     2  node-tests/fixtures/service-test/default-nested.ts:1
     2  node-tests/fixtures/service-test/default.ts:1
```

muziejus commented:
@simonihmig I have been testing this PR against #240, and I've had to add the following dev dependencies in addition to what you have above:

* `@glimmer/component`
* `@types/ember-data`
* `@types/ember-data__adapter`
* `@types/ember-data__model`
* `@types/ember-data__serializer`
* `@types/ember-data__store`

I also had to upgrade `qunit-dom` to 2.0.0 to get types to appear.

muziejus commented:
Thanks, @simonihmig! I think for @chriskrycho's sake, who's had to review #240 like five times now, I'll have that land and then follow up w/ a PR that adds types to the fixtures and adjusts the blueprints to match them--but that doesn't actually have any mechanism for doing the type checks. 

Then, I think, once that PR lands, we can implement the type-checking mechanism you've put together here, so that `yarn lint:blueprints` will only be implemented once the types are in place. 

That makes sense to me, how does it sound to you?
